Bug 1622336 - Trigger schedules API as soon as a push happens

armenzg opened the pull request.

armenzg commented:
@marco-c could you please give me some initial feedback?

marco-c commented on http_service/Dockerfile:
```diff
@@ -10,4 +10,5 @@ RUN pip install --disable-pip-version-check --quiet --no-cache-dir -r /requireme
 
-CMD gunicorn -b 0.0.0.0:$PORT bugbug_http.app --preload --timeout 30 -w 3
+# Run the Pulse listener in the background
+CMD (/code/http_service/listener.py &> /code/http_service/pulse.log &) && gunicorn -b 0.0.0.0:$PORT bugbug_http.app --preload --timeout 30 -w 3
```
We should find a way to log to the foreground rather than a file

armenzg replied:
The whole problem I was facing had to do that in order for new code to be executed I needed to call `docker-compose build bugbug-http-service`.

Redirection to disk was left by mistake. Fixed locally.

marco-c commented on http_service/listener.py:
```diff
@@ -0,0 +1,62 @@
+        rev = body["payload"]["data"]["heads"][0]
+        if branch in ["autoland", "try"]:
+            url = "{}/push/{}/{}/schedules".format(BUGBUG_HTTP_SERVER, branch, rev)
+            response = requests.get(url, headers={"X-Api-Key": "pulse_listener"})
```
I was thinking we'd directly put the job in the queue, but this should work exactly as well.

marco-c commented on http_service/pulse/consumers.py:
```diff
@@ -0,0 +1,48 @@
+        exchange=Exchange("exchange/hgpushes/v2", type="topic", no_declare=True,),
+        routing_key="#",
+        durable=True,
+        # XXX: This should not be auto delete
```
Maybe it makes sense for it to be auto delete? So e.g. if the service is down we don't keep getting messages in the queue

armenzg replied:
If we don't delete the queue then we will keep adding pushes to the Pulse queue and once our Pulse consumer goes back up we will be able to process and catch up.

marco-c replied:
Yep, but we don't want to do that, as those would be old pushes and so probably test selection for them is not needed anymore (and if it is, we'd already have got the request directly via the API).

marco-c commented on http_service/pulse/consumers.py:
```diff
@@ -0,0 +1,48 @@
+
+class ConsumerFactory:
+    @staticmethod
+    def hg_pushes(user, password, callback):
```
I'd make this actually a contextmanager, so the user can do:
```
with ConsumerFactory.hg_pushes(...) as consumer:
    consumer.run()
```

marco-c commented on http_service/pulse/consumers.py:
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env python3
```
Nit: since the amount of code is relatively limited, could you merge the two files?

marco-c commented on http_service/listener.py:
```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/env python3
```
Could you move the code into `http_service/bugbug_http`?
Then we can add the script to `console_scripts` in `http_service/setup.py` so we can run it directly as `bugbug-http-pulse-listener` instead of `/code/http_service/listener.py`

marco-c replied:
(This would also probably avoid the `pulse` addition in `.isort.cfg`).

marco-c commented on http_service/requirements.txt:
```diff
@@ -5,6 +5,7 @@ Flask==1.1.2
 gunicorn==20.0.4
-marshmallow==3.5.2
+kombu==4.6.8
+marshmallow==3.5.1
```
I assume this was unintended

marco-c commented on http_service/pulse/consumers.py:
```diff
@@ -0,0 +1,48 @@
+_CONNECTION_URL = "amqp://{}:{}@pulse.mozilla.org:5671/?ssl=1"
+
+
+def _generate_hg_pushes_queue(user):
```
Nit: you could just inline this function in the `hg_pushes` function below, it's just a line after all (split into multiple because of its length, but still not much code)

armenzg commented:
@marco-c I've rebased and added a commit with a description of the change in case you want to review the [commit message](https://github.com/mozilla/bugbug/pull/1548/commits/c7bed970771efda0f86ba9bfa6f1398a923b4c71).

```
 Bug 1622336 - Initialize schedules service when Hg push happens

This change adds a Pulse listener for Hg pushes as a background process
to `bugbug-http-service`.

This listener works if `PULSE_USER` and `PULSE_PASSWORD` are set before starting
up the Docker containers.

The listener only pays attention to autoland and try pushes. When these happen
we do a GET request from the bugbug deployment. This will get the schedules service
ready for when the Gecko decision task will query it.
```
